alphafold3_pytorch/data/data_pipeline.py

```diff
@@ -175,13 +175,18 @@ def make_msa_features(
 
             polymer_residue_index = -1
 
+            chemtype_constants_cache = {}
+
             for idx, (chemtype, residue_index) in enumerate(
                 zip(chain_chemtype, chain_residue_index)
             ):
                 is_polymer = chemtype < ligand_chemtype_index
                 is_ligand = not is_polymer
 
-                chem_residue_constants = get_residue_constants(res_chem_index=chemtype)
+                if chemtype not in chemtype_constants_cache:
+                    chemtype_constants_cache[chemtype] = get_residue_constants(res_chem_index=chemtype)
```

Comment:
is `get_residue_constants` expensive function?

Reply:
I need to run some more experiments, but caching its execution for each token in a sequence seems to have dropped the runtime of `pdb_input_to_atom_input` from minutes to seconds. Fingers crossed this pattern holds for other structures

Reply:
In principle, it should be a (1) complexity function, since it does a simple lookup based on integers or strings and returned a static Python `Module` class instance.

Reply:
> I need to run some more experiments, but caching its execution for each token in a sequence seems to have dropped the runtime of `pdb_input_to_atom_input` from minutes to seconds. Fingers crossed this pattern holds for other structures

woah 🚀 

and yea, a few lines of caching does wonders, if put in the right place

it can also do a ton of harm if done incorrectly, or not invalidated right lol